cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+will enable us to focus on smaller deliverables that each independently provide
+value to the project.
+
+### 1. Common Interface for Data Sources
+Before we get started on an xDS adapter for Cilium, we'll need to build a common
+interface for interacting with data sources (Kubernetes, KVStore, and xDS). This
+interface will enable the majority of the codebase to be entirely unaware of
+which data sources are in use.
+
+A key part of defining this interface will involve determining a precedence
+order when more than one data source is connected. When any conflicts occur,
+for example if Services from different sources have the same IP, the following
+order will be used for precedence:
+
+1. Kubernetes
+2. KVStore
+3. xDS
+
+If there is interest, we may make this order configurable in a future
+enhancement.
+
+### 2. New xDS Adapter
+In this phase we'll develop a new xDS adapter that implements the interface
+developed in the first phase.
+
+#### API Mapping
+To keep the initial work as focused as possible, I propose using existing xDS
+APIs to model ClusterIP routing. In the future, we may want to consider either
+additions to those existing xDS APIs or new xDS APIs to cover the entirety of
+Cilium functionality. To start, it seems highly valuable to focus on the areas
+where existing xDS APIs overlap with existing Cilium capabilities.
+
+This shows how we could use existing xDS APIs to represent Cilium capabilities
+for ClusterIP routing:
+
+| Cilium Field | Origin xDS | Comments |
+| - | - | - |
+| Frontend.Address | `filter_chain_match.prefix_ranges` | |
+| Frontend.Port | `filter_chain_match.destination_port` | |
+| Frontend.Protocol | `cluster.metadata.protocol` | |
+| Backend[*].FEPortName | `cluster.metadata.port_name` | |
+| Backend[*].NodeName | ? | TBD - This is only used for logic to determine if an endpoint is local. Instead of sending this through xDS for every endpoint we could just look at the CIDR configured for each Cilium instance. |
+| Backend[*].Addr, Backend[*].Port | `lb_endpoints.endpoint.address.socket_address` | |
+| Backend[*].State | `lb_endpoints.health_status` | |
+| Backend[*].Preferred | N/A | Not used in Kubernetes |
+| Backend[*].Weight | `lb_endpoints.endpoint.load_balancing_weight` | |
+| Type | N/A | Out of scope - only focusing on ClusterIP Services |
+| ExternalTrafficPolicy, InternalTrafficPolicy | `cluster.load_balancing_policy` | Defined in the typed struct `type.googleapis.com/cilium_io_traffic_policy` |
+| NatPolicy | N/A | Not used in Kubernetes |
+| SessionAffinity | `tcpproxy.hash_policy` | |
+| SessionAffinityTimeout | ? | TBD - This represents the maximum session stickiness time. There is no natural place for this in existing xDS APIs. |
+| HealthCheckNodePort | N/A | Out of scope - only focusing on ClusterIP Services |
+| Name | `cluster.metadata.service_name` | |
+| Namespace | `cluster.metadata.service_namespace` | |
+
+### 3. Sample xDS Control Plane
```

Comment:
Food for thought: it might make sense to add this xDS Control plane into Cilium Operator.

Reply:
+1 if the community decides to provide this to Cilium users.

Reply:
I'm -1 on adding it into the Operator, personally, because the operator is currently tightly focussed on its current role: reading from Kubernetes objects, building some internal model, then writing back to Kubernetes. The Operator doesn't have any actual routing logic itself.

An xDS control plane will need to have a lot of routing logic added. I think it's better to keep this as a separate thing that's built with the same tooling as the Agent and Operator (that is, using the Hive and Cell pattern), so that it can be merged in later if necessary.

Reply:
Seems reasonable to have a standlone component that implements the control plane. Adding it to the operator has the benefit of not requiring a new component and it can simply be feature gated. Though from a stability perspective, keeping this out of operator is beneficial.